[Graph] Field manager

flash1293 opened the pull request:
## Summary

This PR moves the field manager to react and EUI. It also implements the state handling as redux store.

## Adding new fields

I made it possible to add multiple fields at once because in most cases you want to work with more than one field.
<img width="965" alt="Screenshot 2019-09-20 at 16 05 33" src="https://user-images.githubusercontent.com/1508364/65333304-8ac6d100-dbc0-11e9-9331-3b45c94cf5b0.png">


## Field editor

`Max hops` is moved to a collapsed accordion because in most cases you don't change this setting. If you change the field, the current field is deselected and the new field is selected and gets the settings from the current field (color, icon and hop size). Changes are directly applied (no update button), because they are easy to reverse - even if you remove a field you can just add it again and get all settings back.
<img width="361" alt="Screenshot 2019-09-20 at 16 05 39" src="https://user-images.githubusercontent.com/1508364/65333319-91554880-dbc0-11e9-98a5-a905aa7386b3.png">
<img width="502" alt="Screenshot 2019-09-20 at 16 05 45" src="https://user-images.githubusercontent.com/1508364/65333327-9619fc80-dbc0-11e9-876e-837554f2b144.png">


## Disabled fields

Fields can be disabled by the switch in the field editor or by shift-clicking the field. Not really happy with the style of disabled fields yet.
<img width="633" alt="Screenshot 2019-09-20 at 16 07 12" src="https://user-images.githubusercontent.com/1508364/65333405-bb0e6f80-dbc0-11e9-9063-b2826d59ebe9.png">


## State handling

For now the redux store doesn't do much, but it will be extended in subsequent PRs to do all the state handling currently living in legacy JS files.

flash1293 commented:
Jenkins, test this

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2890/)

flash1293 commented:
@cchaos First take on the field manager. Not super happy with everything, feedback is greatly appreciated.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3028/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3030/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3062/)

elasticmachine commented:
Pinging @elastic/kibana-app

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3381/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3560/)

flash1293 commented:
Jenkins, test this.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3581/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3594/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3636/)

monfera commented:
The field rectangles, which also host the mostly-filled icon squares, feel a sliver too dark. Wondering if a tint around #e8f0fd (same hue but ~10% lighter) would solve it or would be in some way less good
![image](https://user-images.githubusercontent.com/1548516/65165367-ed905f00-da3e-11e9-87d9-82dbf40fcea6.png)

monfera commented:
What's behind the term "Filter options"? It looks like a search field, so maybe "Field search" or "Search"?
![image](https://user-images.githubusercontent.com/1548516/65167546-10bd0d80-da43-11e9-8086-2ea7ac6b5bd1.png)

monfera commented:
Should "Add fields" be inactive when there's no field selected?
![image](https://user-images.githubusercontent.com/1548516/65167690-52e64f00-da43-11e9-89f5-bd25df4420cc.png)

monfera commented:
The toggle also makes the scrollbar run amok (maybe other things too):
![toggle](https://user-images.githubusercontent.com/1548516/65167861-a9538d80-da43-11e9-9b0f-688bff8dc239.gif)

monfera commented:
Just hovering also triggers the vertical scrollbar, and I saw it in other situations, eg. when the field dialog goes away (I'll stop reporting the scrollbar; underlying issue may be the same)
![hover-scrollbar](https://user-images.githubusercontent.com/1548516/65168075-0ea77e80-da44-11e9-894f-26fa73e8320f.gif)

monfera commented:
Not sure what the best way would be for fixing it but there's a somewhat disturbing flash of the help bubble if the user clicks on the field (eg. to configure it):
![flash](https://user-images.githubusercontent.com/1548516/65168342-93929800-da44-11e9-8cb9-176798536cab.gif)

monfera commented:
The help bubble comes back after the user exited the modal; I think it shouldn't, at least it feels unusual to me
![comeback](https://user-images.githubusercontent.com/1548516/65168514-fab04c80-da44-11e9-9a6d-22172e000b40.gif)

monfera commented:
This shows two, likely related issues:

1. The field is somehow in a "selected" status. See the thick blue border around the field, not seen on the others. I don't think there's a notion of the field having this kind of state; it's more of a button, even the dialog isn't on
2. When tabbing back to the Kibana browser tab, the help bubble pops up again and again. The direct cause could well be the above item

![selected](https://user-images.githubusercontent.com/1548516/65169691-54b21180-da47-11e9-96f5-ee88af862274.gif)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3676/)

flash1293 commented:
Thanks for your feedback @monfera 

> The field rectangles, which also host the mostly-filled icon squares, feel a sliver too dark. Wondering if a tint around #e8f0fd (same hue but ~10% lighter) would solve it or would be in some way less good

You are right, that looks better. @cchaos what do you think?

> The scrollbar on the right comes and goes on field hover, maybe because of the bubble appearing atop the field (not part of this gif)?

This only happens for me if I make my browser window super small (200px of height or something like that) Did you test with such a window? It seems like the way the EUI popovers work causes the scrollbar to appear in such extreme conditions.

> What's behind the term "Filter options"? It looks like a search field, so maybe "Field search" or "Search"?

Good point, the shown fields are possible options, and you are filtering them down. It's the default placeholder. I changed it to `Filter fields`.

> Should "Add fields" be inactive when there's no field selected?

It should, fixed.

> Not sure what the best way would be for fixing it but there's a somewhat disturbing flash of the help bubble if the user clicks on the field (eg. to configure it):

Seems like an EUI issue if you have both a popover and a tooltip on a button. The popover should open to the bottom and then I don't see this flash. It seems like you also tested with a very small browser window here. This shouldn't occur with normal window sizes.

> The help bubble comes back after the user exited the modal; I think it shouldn't, at least it feels unusual to me

Hm, I see what you mean. The popover gives back the focus to the button when it closes, which is probably the right thing to do.

> The field is somehow in a "selected" status. See the thick blue border around the field, not seen on the others. I don't think there's a notion of the field having this kind of state; it's more of a button, even the dialog isn't on

Yes, it is a button element and behaves like a button - you can tab through them and if you click the button, the popover opens. I think this is the best behavior for keyboard users.

monfera commented:
@flash1293 thanks for the changes and the answers! Replying to only what's not resolved:

> This only happens for me if I make my browser window super small (200px of height or something like that) Did you test with such a window? It seems like the way the EUI popovers work causes the scrollbar to appear in such extreme conditions.

and

> It seems like you also tested with a very small browser window here. This shouldn't occur with normal window sizes

These were fullscreen, I think the gif also shows they weren't small

> It seems like the way the EUI popovers work causes the scrollbar to appear in such extreme conditions. 

It's weird but it wasn't sync with the bubble; for every _cycle_ of the bubble, it toggled on, or off. So, related, but not apparently caused by it.

> [your last two responses with the field focus issue]

I see that a concept focus is needed for accessibility. But I think these field symbols should work more like eg. links, at least it feels annoying that the help bubble reappears on dialog exit or switching back to the Kibana browser tab, and even the box around the field feels futile, as it's otherwise a stateless thing (again, except for accessibility). What do you think, @cchaos, maybe it's an EUI-wide decision choice?

monfera commented:
Thanks Joe, your latest commit resolved all the scrollbar and bubble issues, so the only remaining question is for @cchaos (of lightening the bluish grey fill)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3836/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3845/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3853/)

cchaos commented:
@flash1293 Sorry it took me so long, priorities and all that. But here is a prototype for you: https://marvelapp.com/b25a7g7/screen/61591740

I like the idea of being able to quickly add multiple fields, but that's definitely something we should confirm with users via testing/feedback. The display of the fields should really just be EuiBadges and the edit menu should behave similarly to the global filters. An established pattern they can remember throughout apps.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4096/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4105/)

flash1293 commented:
@cchaos I like it being consistent with filters - I updated the PR, could you take another look? While I was at it, I changed the default field colors to use the light colors of the eui palette. I have the same problem as maps that combo box can't show an icon as selected option.

cchaos commented:
Design PR4U: https://github.com/flash1293/kibana/pull/4

cchaos commented:
> I have the same problem as maps that combo box can't show an icon as selected option.

That's fine for now, I'll look for or create a ticket for this type of display in EUI.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4114/)

cchaos commented:
I also noticed that it seemed like the field edits were not trickling down to the Graph itself?

<img src="https://d.pr/free/i/HFsZEN+" width="50%" />
<img src="https://d.pr/free/i/ajSTxb+" width="50%" />

flash1293 commented:
@cchaos the disabling thing is expected, disabling the field just prevents new nodes of this type to be discovered via searches and expands. The color should get updated, good catch - I will look into this.

flash1293 commented:
@cchaos I couldn't reproduce the problem with settings not trickling down to the graph - it's working fine for me. Could you check whether it's reproducible?
![change_field_settings](https://user-images.githubusercontent.com/1508364/65410792-8b8f7b00-ddeb-11e9-9be0-319817d91d1d.gif)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/213/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/228/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/241/)

cchaos commented:
@flash1293 Hmmm, I pulled down the lates and I'm not having that issue anymore. I wonder if it was because I was working off an already created Graph (in a different version) but since I've started new, it's been working. I'll keep an eye out for if it happens again.

cchaos commented on x-pack/legacy/plugins/graph/public/components/field_manager/field_icon.tsx:
```diff
@@ -0,0 +1,37 @@
+}
+
+function getIconForDataType(dataType: string) {
+  const icons: Partial<Record<string, UnwrapArray<typeof ICON_TYPES>>> = {
```
Seems like you're doing this same thing in Lens. Maybe you'll want to figure out a more global service?

cchaos commented on x-pack/legacy/plugins/graph/public/components/field_manager/field_icon.tsx:
```diff
@@ -0,0 +1,37 @@
+export function getColorForDataType(type: string) {
+  const iconType = getIconForDataType(type);
+  const { colors } = palettes.euiPaletteColorBlind;
+  const colorIndex = stringToNum(iconType) % colors.length;
```
We should definitely make sure we're using the same colors across all apps.
